### ISSUE TITLE
actions executed twice in cy.wrap()

### ISSUE THREAD

**Issue by agentHoover** [cypress-io/cypress#2333]
### Current behavior:
In order to handle a custom control checkbox I was playing around with overwriting the default check method.    

_The basic problem I am trying to overcome is that the check method against the actual input element fails because cypress sees the parent ```<div class="mat-checkbox-inner-container>``` element as covering the input control  -- hence the need for the force parameter in the custom command._ 

The html the custom control generates looks like this:
```html
<mat-checkbox class="mat-checkbox mat-accent ng-untouched ng-pristine ng-valid" name="isFlyout" ng-reflect-disabled="false" ng-reflect-id="isFlyout_8" ng-reflect-name="isFlyout" ng-reflect-is-disabled="false" ng-reflect-model="false" id="isFlyout_8">
   <label class="mat-checkbox-layout" for="isFlyout_8-input">
      <div class="mat-checkbox-inner-container mat-checkbox-inner-container-no-side-margin">
        <input class="mat-checkbox-input cdk-visually-hidden" type="checkbox" id="isFlyout_8-input" name="isFlyout" tabindex="0" aria-label="" aria-checked="false">
      <div class="mat-checkbox-ripple mat-ripple" matripple="" ng-reflect-centered="true" ng-reflect-radius="25" ng-reflect-animation="[object Object]" ng-reflect-disabled="false" ng-reflect-trigger="[object HTMLLabelElement]"></div>
      <div class="mat-checkbox-frame"></div><div class="mat-checkbox-background">
      <svg xml:space="preserve" class="mat-checkbox-checkmark" focusable="false" version="1.1" viewBox="0 0 24 24"><path class="mat-checkbox-checkmark-path" d="M4.1,12.7 9,17.6 20.3,6.3" fill="none" stroke="white"></path>
      </svg>
      <div class="mat-checkbox-mixedmark">/div>
</div>
</div>
<span class="mat-checkbox-label">
<span style="display:none">&nbsp;</span>
</span></label></mat-checkbox>

```
Here is the overwrite command.
```js
Cypress.Commands.overwrite('check', (originalfn, parentDiv) => {
    cy
        .wrap(parentDiv)
        .find('input')
        .check({ force: true });

});

//used like this
cy.get('#isFlyout_8').check();
//where #isFlyout_8 is the parent mat-checkbox element of the control

```

When this runs the wrap/find on the initial parentDiv control correctly finds the child input element but then it wraps the child element and executes the find again. 
![doublewraplog](https://user-images.githubusercontent.com/4604810/44056447-e317cb52-9f0d-11e8-97e8-f29c85a995be.PNG)

### Desired behavior:
The second wrap/find should not be executed. 

### Versions

3.0.3 running on Windows 10 against Chrome.

**Comment by brian-mann** [comment]
You're not returning anything from the custom command, so Cypress has no idea to await what you've provided. You need to return `cy.wrap(...).find(...).check(...)` so that it's aware of what you expect to return. As is, `undefined` is returned.

This may not actually be the root cause, but I'm sure that's not helping. Try that and see if that changes anything.

**Comment by agentHoover** [comment]
I modified the command to: 
```
Cypress.Commands.overwrite('check', (originalfn, checkbox) => {
    return cy
        .wrap(checkbox)
        .find('input')
        .check({ force: true });

});
```

It did not change the result. Still getting a double execution of the commands.

**Comment by lukeapage** [comment]
is this because you use check in the check overwrite, instead of using originalfn...

**Comment by jennifer-shehane** [comment]
@agentHoover Yes, please try to update as @lukeapage suggests and let us know if this is resolved. Like:

```js
Cypress.Commands.overwrite('check', (originalFn, checkbox, name, options) => {
  options.force = true

  return cy
    .wrap(checkbox)
    .find('input')
    .then(($el) => {
      return originalFn($el, name, options)
    })
});
```

**Comment by agentHoover** [comment]
I haven't touched this code in a couple of months. I'm on a different project now.  I will try to get back to this and test it this weekend.

**Comment by benmosher** [comment]
Hi there, I am seeing this with regular `add` with the following command definition:

```
/**
 * see https://github.com/cypress-io/cypress/issues/299#issuecomment-380197761
 * and the hosting issue for why this exists
 */
Cypress.Commands.add(
  'pressTab',
  { prevSubject: 'element' },
  (subject, { event = 'keydown', ...options } = {}) => {
    return cy.wrap(subject).trigger(event, {
      keyCode: 9,
      which: 9,
      ...options,
    })
  },
)
```

and the leveraging test:
```
       it('tab-completes', function() {
          cy.maxbox().pressTab({ force: true })

          cy.get(MAXBOX_PLACEHOLDER).should('not.exist')
          cy.get(MAXBOX_LINE).should(
            'have.text',
            'sales by year for region = northwest',
          )
        })
```

When executing, it runs once at the correct time, before my assertions. They pass, then it looks like the whole test starts to run a second time: the trigger runs again, and then I see a failed assertion in the console/log: 

<img width="577" alt="image" src="https://user-images.githubusercontent.com/3709196/52793141-06650c00-303b-11e9-8a21-6db1d5f6b4c4.png">

(note also that the test succeeds despite the failed assertion on the second attempt?) 

vs this when I replace the custom command with a direct `trigger`, which is what I would expect:

<img width="578" alt="screen shot 2019-02-14 at 09 26 59" src="https://user-images.githubusercontent.com/3709196/52793041-cb62d880-303a-11e9-98f2-0beaf3d107b5.png">

If I don't `force: true` (needed for CodeMirror) I still see two events and actually got a false-positive test success where it just ignored my assertions completely:

<img width="512" alt="image" src="https://user-images.githubusercontent.com/3709196/52793784-817af200-303c-11e9-9dcf-740a2aa939f9.png">

FWIW, I am suspicious maybe something is up with `wrap`? I reduced my command to just `cy.wrap(subject)` and moved the trigger into the test and I still see the double-execution:

<img width="573" alt="image" src="https://user-images.githubusercontent.com/3709196/52794129-39100400-303d-11e9-83fd-0fe67f947e34.png">

**Comment by benmosher** [comment]
Ah, okay, one more piece of info further condemning `wrap`: adding a meaningless `wrap` to my test also double fires:

```
          cy.maxbox().then($el =>
            cy.wrap($el).trigger('keydown', {
              keyCode: 9,
              which: 9,
              force: true,
            }),
          )
```
Maybe `wrap`'s retry implementation is not registering the first pass's success somehow?

Anyway, I can work around by just avoiding `wrap` and directly triggering but it does make defining custom child commands a non-starter, I think.

**Comment by jennifer-shehane** [comment]
@benmosher 	I'm not able to reproduce the `wrap` causing the trigger to happen twice in my example. 

test code
```js
it('wrap with action', () => {
  cy.visit('http://localhost:8080/commands/actions')
  cy.get('#email1').then(($el) => {
    cy.wrap($el)
      .trigger('keydown', {
        keyCode: 9,
        which: 9,
        force: true,
      })
  })
})
```

app code
```
  $('#email1').on('keydown', () => {
    console.log('keydown')
  })
```

console log
<img width="497" alt="screen shot 2019-02-19 at 1 36 57 pm" src="https://user-images.githubusercontent.com/1271364/52996312-9d6f0100-344b-11e9-8f2e-fae5cc45044e.png">

**Comment by benmosher** [comment]
hmmm, maybe the custom command is part of it too? I will try to get a super-minimal repro if I can, but here is my `maxbox`command implementation:

```
Cypress.Commands.add('maxbox', text => {
  const maxbox = cy.get('[data-test=maxbox] .CodeMirror-wrap textarea')

  if (text) return maxbox.type(text, { force: true, timeout: LOAD_TIME })
  else return maxbox
})```

**Comment by jennifer-shehane** [comment]
Unfortunately we have to close this issue as there is not enough information to reproduce the problem. This does not mean that your issue is not happening - it just means that we do not have a path to move forward. 

Please comment in this issue with a reproducible example and we will reopen the issue. 🙏

**Comment by nnevala** [comment]
I've just now run into this issue myself, which is perfect, seeing how the GH issue was closed only a few hours ago 😄 

I'm running Cypress 3.3.1 on a Mac, and the one bit of additional detail I can offer is that in my case I've only been able to reproduce the problem when running Cypress in a headless mode. And even then it doesn't seem to always happen (maybe 4 times out of 5).

Otherwise my case matches the other findings in this thread: a `wrap` call on a file input element in a custom command causes the subsequent `change` event to be triggered twice.
